tick/stateful/functions.go

```diff
@@ -12,21 +12,76 @@ import (
 
 	"github.com/dustin/go-humanize"
 	"github.com/influxdata/influxdb/influxql"
+	"github.com/influxdata/kapacitor/tick/ast"
 )
 
+const (
+	maxArgs = 4
+)
+
+type ErrWrongFuncSignature struct {
+	Name           string
+	DomainProvided Domain
+	Func           Func
+}
+
+func (e ErrWrongFuncSignature) Error() string {
+	return fmt.Sprintf("Cannot call function \"%s\" with args %s, available signatures are %s.",
+		e.Name, e.DomainProvided, FuncDomains(e.Func))
+}
+
 var ErrNotFloat = errors.New("value is not a float")
 
+type Domain [maxArgs]ast.ValueType
+
+func (d Domain) String() string {
+	input := []string{}
+	for _, el := range d {
+		if el == ast.InvalidType {
+			// Because inputs should be consecutive
+			break
+		}
+		input = append(input, el.String())
+	}
+
+	return "(" + strings.Join(input, ",") + ")"
+}
+
+type Domains []Domain
+
+func (ds Domains) String() string {
+	input := []string{}
+	for _, d := range ds {
+		input = append(input, d.String())
+	}
+
+	return "[" + strings.Join(input, ", ") + "]"
+}
+
 // A callable function from within the expression
 type Func interface {
 	Reset()
 	Call(...interface{}) (interface{}, error)
+	Signature() map[Domain]ast.ValueType
+}
+
+func FuncDomains(f Func) Domains {
+	ds := []Domain{}
+
+	for d := range f.Signature() {
+		ds = append(ds, d)
+	}
+
+	return ds
 }
 
 // Lookup for functions
 type Funcs map[string]Func
 
 var statelessFuncs Funcs
 
+var GlobalExecutionSate ExecutionState
```

Comment:
As far as I can tell this is just used to lookup the built in functions as opposed to the dynamic functions. And only the map of functions is used. Maybe this would be better as

```
var builtinFunc Func
```

Also I think it should be private since it is not used outside the package.